### PR TITLE
Refactor: move assertions to start of functions

### LINKED ISSUE #537: Move the assertions to the beginning of function
It is desirable to have assertions in the beginning of a function because the compiler might be able to improve its optimization, and it is more readable for the users to understand the panic conditions of a function

_Originally posted by @bobbinth in https://github.com/maticnetwork/miden/pull/530#discussion_r1028397476_

### PR DESCRIPTION
**Pull request by MuhtasimTanmoy**
## Describe your changes
This PR moves assertions to the start of functions. Resolves #537.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to the naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

**Comment by MuhtasimTanmoy** [comment]
[Rustfmt](https://github.com/rust-lang/rustfmt) and [Clippy](https://github.com/rust-lang/rust-clippy) linting provided two change requests.

```
- assert_eq!(expected, format!("{}", program));  // change this to the following
+ assert_eq!(expected, format!("{program}"));
```

And,

```
-  assert_eq!(expected, format!("{}", program));  // change this to the following
+  assert_eq!(expected, format!("{program}"));
```

Do I need to update those occurrences too?

**Comment by MuhtasimTanmoy** [comment]
Should I change the numeric value in the following line to a `CONSTANT`  declared in [assembly/src/lib.rs](https://github.com/0xPolygonMiden/miden-vm/blob/main/assembly/src/lib.rs)

```rust
assert_eq!(self.module_stack.len(), 1, "executable not last module. module_stack len should be 1");
```

to

```rust
assert_eq!(self.module_stack.len(), SOME_CONSTANT, "executable not last module. module_stack len should be 1");
```
this?

File: assembly/src/assembler/context.rs [hunk 1]
```diff
@@ -439,14 +438,17 @@ impl ModuleContext {
...
+    /// - If this module is not an executable module.
+    /// - If the procedure context is not main.
     pub fn complete_executable(&mut self) {
         assert!(self.is_executable(), "module not executable");
-
-        let main_proc_context = self.proc_stack.pop().expect("no procedures");
-
+        assert_eq!(self.proc_stack.len(), 1, "procedure stack should only have the main procedure. proc_stack length should be 1");
+        let main_proc_context = self.proc_stack.pop().unwrap();
         assert!(main_proc_context.is_main(), "not main procedure");
```
Comment (vlopes11) [inline assembly/src/assembler/context.rs:451]: I wonder if this shouldn't be a case to justify a signature change for the function...Usually, when we can't describe the assertion at the beginning of the function, then chances are we the function shouldn't be infallible.

But, sticking to the scope of the PR, the implementation is fine as it is!

File: assembly/src/assembler/context.rs [hunk 2]
```diff
@@ -439,14 +438,17 @@ impl ModuleContext {
...
-        assert!(self.proc_stack.is_empty(), "more procedures after main");
```
Comment (vlopes11) [inline assembly/src/assembler/context.rs:448]: Good catch! If the length is asserted, then this check is redundant.

**Review by vlopes11 — CHANGES_REQUESTED** [review]
The changes looks good! We shouldn't close the issue yet because we need to check the other places as well.

However, the CI is breaking with `rustfmt`. Be sure to execute `cargo fmt` so it will pass on that test.

**Comment by MuhtasimTanmoy** [comment]
> Be sure to execute cargo fmt so it will pass on that test.

Done.

File: assembly/src/assembler/context.rs [hunk 3]
```diff
@@ -439,14 +442,21 @@ impl ModuleContext {
     /// compiling a program, the executable module will have the main procedure left on the
     /// procedure stack. To complete the module we need to pop the main procedure off the stack and
     /// append its callset to the callset of the module context.
+    ///
+    /// # Panics
+    /// Panics if:
+    /// - There is not exactly one module left on the procedure stack.
+    /// - If this module is not an executable module.
+    /// - If the procedure context is not main.
```
Comment (bobbinth) [inline assembly/src/assembler/context.rs:450]: nit: I'd change these to:
```
- If this module is not an executable module.
- If there is not exactly one procedure left on the procedure stack.
- If the procedure left on the procedure stack is not main procedure.
```

File: assembly/src/assembler/context.rs [hunk 4]
```diff
@@ -439,14 +442,21 @@ impl ModuleContext {
...
     pub fn complete_executable(&mut self) {
         assert!(self.is_executable(), "module not executable");
-
-        let main_proc_context = self.proc_stack.pop().expect("no procedures");
-
+        assert_eq!(
+            self.proc_stack.len(),
+            1,
+            "procedure stack should only have the main procedure. proc_stack length should be 1"
```
Comment (bobbinth) [inline assembly/src/assembler/context.rs:456]: nit: I'd change the message to: "procedure stack must contain exactly one procedure".

File: assembly/src/assembler/context.rs [hunk 5]
```diff
@@ -225,9 +225,12 @@ impl AssemblyContext {
     ///   procedure cache or the local procedure set of the module.
     pub fn into_cb_table(mut self, proc_cache: &ProcedureCache) -> CodeBlockTable {
         // get the last module off the module stack
-        let mut main_module_context = self.module_stack.pop().expect("no modules");
-        assert!(self.module_stack.is_empty(), "executable not last module");
-
+        assert_eq!(
+            self.module_stack.len(),
+            1,
+            "executable not last module. module_stack len should be 1"
```
Comment (bobbinth) [inline assembly/src/assembler/context.rs:231]: nit: I'd change the message to: "module stack must contain exactly one module".

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you for working on this! I think it looks good - I left just a couple of nits inline. To answer your questions:

> Should I change the numeric value in the following line to a `CONSTANT` declared in [assembly/src/lib.rs](https://github.com/0xPolygonMiden/miden-vm/blob/main/assembly/src/lib.rs)

I don't think this is needed.

> [Rustfmt](https://github.com/rust-lang/rustfmt) and [Clippy](https://github.com/rust-lang/rust-clippy) linting provided two change requests.
> ...
> Do I need to update those occurrences too?

Would be great, but not required.

**Comment by MuhtasimTanmoy** [comment]
Fix given as per feedback.

**Comment by bobbinth** [comment]
@MuhtasimTanmoy - looks good! Could you fix the `rustfmt` issue and we can merge.

**Review by bobbinth — APPROVED** [review]

**Comment by MuhtasimTanmoy** [comment]
> Could you fix the rustfmt issue and we can merge.

Done as suggested.

------

However, there is something that has left me perplexed which I would like to bring to your attention.

In our previous workflow run, `cargo fmt` suggested converting this line
```
assert_eq!(self.module_stack.len(), 1, "executable not last module. module_stack len should be 1");
```
to

```
assert_eq!(
            self.module_stack.len(),
            1,
            "module stack must contain exactly one module"
        );
```
So, as suggested, I performed this action in commit https://github.com/0xPolygonMiden/miden-vm/pull/707/commits/3639582e5d9e7c648231af38cfc2b1f893c5bea1, which eliminated all errors from the workflow check.


However, in our last workflow run, the error highlighted by "cargo fmt" suggested converting this line
```
assert_eq!(
            self.module_stack.len(),
            1,
            "module stack must contain exactly one module"
        );
```
back into
```
assert_eq!(self.module_stack.len(), 1, "executable not last module. module_stack len should be 1");
```

So, I reverted back to the previous structure in my last commit https://github.com/0xPolygonMiden/miden-vm/pull/707/commits/70bcf4abb47e6b5bde2cd7e2311cfc5dee87b814

I don't particularly get the reason for this.

**Pull request merged**